### PR TITLE
Enables ability to create multiple versions of the builder

### LINKED ISSUE #2: Support multiple versions of Phoenix

### PR DESCRIPTION
**Pull request by jtslear**
- closes openshift-s2i/s2i-phoenix#2
- Breaks out the build environments including the Dockerfiles, the bin
  scripts, and the small test app into respective versions
- Utilizes the same manner openshift is going about using make files and
  shell scripts to perform multi version builds
- Removes the hacky solution for the nodejs install and does it properly
- Uses the latest version for the variety software packages
- Enables the use of buildargs incase one may want to override them

### CODE REVIEW

File: hack/build.sh [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  DOCKERFILE_PATH=$(perl -MCwd -e 'print Cwd::abs_path shift' $dockerfile)
+  cp ${DOCKERFILE_PATH} "${DOCKERFILE_PATH}.version"
+  git_version=$(git rev-parse HEAD)
+  echo "LABEL io.openshift.builder-version=\"${git_version}\"" >> "${dockerfile}.version"
+  if [[ "${UPDATE_BASE}" == "1" ]]; then
+    BUILD_OPTIONS+=" --pull=true"
+  fi
+  docker build ${BUILD_OPTIONS} -t ${IMAGE_NAME} -f "${dockerfile}.version" .
+  if [[ "${SKIP_SQUASH}" != "1" ]]; then
+    squash "${dockerfile}.version"
```
Comment (maennchen) [inline hack/build.sh:40]: What does this `squash` command do? Where can I get it?

Could we point to the `squash` binary in the `README`?

Comment (jtslear) [reply hack/build.sh:40]: You know, that is a great question.  I pulled this from other s2i's that have the squash command available.  I've never actually ran it with this option....

**Comment by jtslear** [comment]
@jawnsy I realize this PR is ridiculously large, but it enables being able to bring online and test multiple version of the Phoenix web framework.  What can I do to help get this PR pushed across the finish line?

File: hack/build.sh (hunk 1 above)
Comment (jawnsy) [reply hack/build.sh:40]: http://jasonwilder.com/blog/2014/08/19/squashing-docker-images/

The squash command should be installed automatically using easy_install, but I guess the function declaration was removed from this image: https://github.com/jawnsy/s2i-go/blob/master/hack/build.sh#L60

File: 1.2/test/run [hunk 2]
```diff
@@ -27,11 +27,12 @@ container_exists() {
 }
 
 container_ip() {
-  if [ -n "$DOCKER_MACHINE_NAME" ]; then
-    docker-machine ip ${DOCKER_MACHINE_NAME}
-  else
-    docker inspect --format="{{ .NetworkSettings.IPAddress }}" $(cat $cid_file)
-  fi
+  #if [ -n "$DOCKER_MACHINE_NAME" ]; then
```
Comment (jawnsy) [inline 1.2/test/run:30]: Was this intended?

File: hack/build.sh (hunk 1 above)
Comment (bparees) [reply hack/build.sh:40]: it's also acceptable to get rid of the squash logic entirely.

File: 1.2/test/run [hunk 3]
```diff
@@ -157,9 +158,9 @@ test_application() {
 
   test_scl_usage "elixir --version" "OTP 18"
   check_result $?
-  test_scl_usage "elixir --version" "Elixir 1.3.1"
+  test_scl_usage "elixir --version" "Elixir 1.2.5"
```
Comment (jawnsy) [inline 1.2/test/run:161]: Curious why Elixir was downgraded?

**Comment by jawnsy** [comment]
@jtslear I took a brief look through, and not knowing anything about Phoenix or Elixir, these changes seem reasonable to me. I'm assuming you tested this stuff locally? If that's the case, then I suggest just merge this

File: 1.2/test/run (hunk 3 above)
Comment (maennchen) [reply 1.2/test/run:161]: It is only downgraded in the `1.2` folder. There are other directories for version `1.3` and possibly `1.4` in the future. jtslear/s2i-phoenix#1

**Comment by maennchen** [comment]
@jawnsy I also tested `1.3` locally. It seems good to me.

**Review by maennchen — APPROVED** [review]

File: 1.2/test/run (hunk 3 above)
Comment (jtslear) [reply 1.2/test/run:161]: Since I started with the newest version, my only method to test multiple version was to introduce a downgraded version :-P

**Comment by jtslear** [comment]
If I had merge privileges I would :-(

**Comment by maennchen** [comment]
@jtslear Did you see my PR on you repo btw? Is it ok?

(I could also wait for the merge of this PR and rebase it.)

**Comment by jtslear** [comment]
@maennchen sorry I missed it.  Throw a PR into this repo.  Sorry about that.

**Comment by maennchen** [comment]
@jtslear Already done, no worries.

**Pull request merged**